src/harness/reference_models/fss_purge/fss_purge.py

```diff
@@ -0,0 +1,300 @@
+
+    if distance_km <= 40 and cbsd['grants']:
+        neighboring_cbsds_with_grants.append(cbsd)
+  return neighboring_cbsds_with_grants
+
+def performPurge(cbsds, fss_entity):
+  """ Implementation of algorithm described in R2-SGN-29.
+  The list of grants to be purged based on the interference caused by their
+  out of band emission are identified and the grants are removed from the CBSDs.
+
+  Args:
+    cbsds: List of CbsdData objects that are in the neighborhood to an FSS.
+    fss_entity: The FssProtectionPoint tuple of the FSS for which the purge is to be performed.
+  """
+
+  # Calculate the OOBE interference value for each CBSD based on the grant with the highest frequency.
+  calculateOobeInterference(cbsds, fss_entity)
+
+  # Sort the CBSDs based on the calculated interference value.
+  sorted_cbsds = sorted(cbsds, key=lambda cbsd: cbsd['oobe_interference'])
+  # Calculate the threshold value in dBm for the OOBE interference
+  oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
+
+  # Find the largest index such that aggregation of the interferences do not cross 
+  # the OOBE threshold
+  oobe_interferences = [interference.dbToLinear(cbsd['oobe_interference']) for cbsd in sorted_cbsds]
+  cumulated_interference = np.cumsum(oobe_interferences)
+  index = np.searchsorted(cumulated_interference, interference.dbToLinear(oobe_threshold_value))
+  cbsds_to_purge = sorted_cbsds[index:]
+  sorted_cbsds = sorted_cbsds[:index]
+  # Remove the grants from the CBSDs that are in the purge list.
+  if cbsds_to_purge:
+    for index, cbsd in enumerate(cbsds_to_purge):
+      if cbsd['grants']:
+        # Get the grant that was used for the OOBE interference calculation, that
+        # is the grant with highest highFrequency.
+        grant_with_max_high_freq = max(cbsd['grants'],\
+              key=lambda x: x['operationParam']['operationFrequencyRange']['highFrequency'])
+
+        # Get all the grants that share the same emission mask as the one that 
+        # was used for OOBE calculation
+        if (grant_with_max_high_freq['operationParam']['operationFrequencyRange']
+            ['highFrequency'] > 3690000000):
+
+          # Adding the -13 emission mask grants to the purge list 
+          
+          grants_to_purge = [grant for grant in cbsd['grants'] 
+                              if grant['operationParam']['operationFrequencyRange']
+                              ['highFrequency'] > 3690000000]
+        else:
+          # Adding the -25 emission mask grants to the purge list
+          grants_to_purge = [grant for grant in cbsd['grants'] 
+                              if grant['operationParam']['operationFrequencyRange']
+                              ['highFrequency'] <= 3690000000]
+
+        # Removing grant requests that are in the purge list from the CBSDs.
+        for grant_items in grants_to_purge:
+           cbsd['grants'].remove(grant_items)
```

Comment:
Each FSS must be considered independently for all grants. Here the grants are being removed from the input which can result in FSS purge list results being dependent on the order the FSSs are provided in. Please refactor this to prevent the result being FSS order dependent.

Reply:
Good point about grants interfering with more than one FSS. However, I think that the dependency has to be taken into account and the grants removed for one FSS should not be considered for another FSS. A grant removed for one FSS should not be present in the final list and I think it should also not be considered for another FSS because considering that grant might cause some other valid grant to get purged.

For example consider two FSSs F1, F2 and six Grants G1, G2, G3, G4, G5 and G6. If the grants_to_purge for ‘F1’ is [G4, G5] and if these two grants are considered for F2 and the sorted list of grants_to_keep becomes [G4,G5, G3, G1] and the grants_to_purge for F2 becomes [G2, G6] then grant G2 and G6 would get purged. If we had not considered the grants G4 and G5 then perhaps grant G2 or both grants G2 and G6 might have ended in grants_to_keep list and would not get purged.

I completely agree on the point that the order of FSSs should not be a dependency. I propose an approach wherein the FSS purge model for all FSS records is executed recursively if the previous run resulted in any grants getting purged for any of the FSS records. A high level psedo-code for the approach is below.

fssPurgeReferenceModel (uut_fad, sas_th_fads, fss_records)
  set cbsd_list to the list of all cbsds from the  fad objects of UUT and THs
  re_run_flag = False
  for each fss_record in fss_records
    no_of_grants_purged = performPurge(cbsd_list, fss_record)
    if no_of_grants_purged > 0
      re_run_flag = True

  if re_run_flag:
    fssPurgeReferenceModel(uut_fad, sas_th_fads, fss_records)

Reply:
This doesn't match the spec. I suggest you talk to Masoud and Sam. After that if you still think it is appropriate please bring it up in the Tuesday call.

Reply:
Taran

After discussing with Sam, we believe what you suggested is correct and thus we make the changes to reflect this.